File: TCN/PythonModel/inference.py
"""
TinyTCN INT8 inference script (PTQ version)

Supports two inference paths:
1. Use the PyTorch quantized model (.pth)
2. Use exported binary weights (.bin) for manual INT8 inference

Use this to verify both paths produce consistent results.
"""

import struct
import numpy as np
import torch
import torch.nn as nn
import torch.ao.quantization as tq
from model import TinyTCNInt8

# ==========================================
#               Configuration
# ==========================================
BIN_PATH = "../outputs/tiny_tcn_int8.bin"
PTH_PATH = "../outputs/tiny_tcn_int8.pth"
DATA_PATH = "../../pdn_dataset_uint8.npz"

# Model structure parameters (must match model.py)
NUM_INPUTS = 9
NUM_CHANNELS = [11, 10, 5, 4]
KERNEL_SIZE = 5
WINDOW_SIZE = 50


# ==========================================
#           Binary file parser
# ==========================================
class BinaryModelReader:
    """Parse .bin files exported by export_flat_binary"""
    
    def __init__(self, filepath: str):
        with open(filepath, 'rb') as f:
            self.data = f.read()
        self.offset = 0
        
    def read_f32(self) -> float:
        val = struct.unpack('f', self.data[self.offset:self.offset+4])[0]
        self.offset += 4
        return val
    
    def read_i32(self) -> int:
        val = struct.unpack('i', self.data[self.offset:self.offset+4])[0]
        self.offset += 4
        return val
    
    def read_f32_array(self, count: int) -> np.ndarray:
        vals = struct.unpack(f'{count}f', self.data[self.offset:self.offset+4*count])
        self.offset += 4 * count
        return np.array(vals, dtype=np.float32)
    
    def read_i32_array(self, count: int) -> np.ndarray:
        vals = struct.unpack(f'{count}i', self.data[self.offset:self.offset+4*count])
        self.offset += 4 * count
        return np.array(vals, dtype=np.int32)
    
    def read_i8_array(self, count: int) -> np.ndarray:
        vals = struct.unpack(f'{count}b', self.data[self.offset:self.offset+count])
        self.offset += count
        return np.array(vals, dtype=np.int8)


class LayerParams:
    """Store per-layer quantization parameters and weights"""
    def __init__(self):
        self.input_scale = 0.0
        self.input_zp = 0
        self.output_scale = 0.0
        self.output_zp = 0
        self.weight_scales = None  # per-channel scales
        self.weight_zps = None     # per-channel zero points
        self.bias_int32 = None
        self.weights_int8 = None
        self.weight_shape = None


def read_layer_params(reader: BinaryModelReader, 
                      out_channels: int, 
                      weight_shape: tuple) -> LayerParams:
    """Read quantization parameters and weights for one layer"""
    params = LayerParams()
    
    # Quantization parameters
    params.input_scale = reader.read_f32()
    params.input_zp = reader.read_i32()
    params.output_scale = reader.read_f32()
    params.output_zp = reader.read_i32()
    
    # Weight parameters (per-channel)
    params.weight_scales = reader.read_f32_array(out_channels)
    params.weight_zps = reader.read_i32_array(out_channels)
    params.bias_int32 = reader.read_i32_array(out_channels)
    
    # INT8 weights
    weight_count = np.prod(weight_shape)
    params.weights_int8 = reader.read_i8_array(weight_count).reshape(weight_shape)
    params.weight_shape = weight_shape
    
    return params


def load_binary_weights(filepath: str):
    """
    Load all weights from the .bin file
    
    Return structure:
    {
        'blocks': [
            {
                'conv1': LayerParams,
                'conv2': LayerParams,
                'downsample': LayerParams or None,
                'block_out_scale': float,
                'block_out_zp': int
            },
            ...
        ],
        'head': LayerParams
    }
    """
    reader = BinaryModelReader(filepath)
    model_weights = {'blocks': [], 'head': None}
    
    # Block structure definition
    block_configs = [
        {'in_ch': NUM_INPUTS, 'out_ch': NUM_CHANNELS[0], 'dilation': 1},
        {'in_ch': NUM_CHANNELS[0], 'out_ch': NUM_CHANNELS[1], 'dilation': 2},
        {'in_ch': NUM_CHANNELS[1], 'out_ch': NUM_CHANNELS[2], 'dilation': 4},
        {'in_ch': NUM_CHANNELS[2], 'out_ch': NUM_CHANNELS[3], 'dilation': 8},
    ]
    
    for i, cfg in enumerate(block_configs):
        in_ch = cfg['in_ch']
        out_ch = cfg['out_ch']
        
        block = {}
        
        # Conv1: in_ch -> out_ch
        conv1_shape = (out_ch, in_ch, KERNEL_SIZE)
        block['conv1'] = read_layer_params(reader, out_ch, conv1_shape)
        
        # Conv2: out_ch -> out_ch
        conv2_shape = (out_ch, out_ch, KERNEL_SIZE)
        block['conv2'] = read_layer_params(reader, out_ch, conv2_shape)
        
        # Downsample: in_ch -> out_ch (1x1 conv)
        if in_ch != out_ch:
            ds_shape = (out_ch, in_ch, 1)
            block['downsample'] = read_layer_params(reader, out_ch, ds_shape)
        else:
            block['downsample'] = None
        
        # Block output scale/zp (for residual add)
        block['block_out_scale'] = reader.read_f32()
        block['block_out_zp'] = reader.read_i32()
        
        model_weights['blocks'].append(block)
    
    # Head: Linear(4, 1)
    head_shape = (1, NUM_CHANNELS[-1])
    model_weights['head'] = read_layer_params(reader, 1, head_shape)
    
    print(f"[Binary] Read complete, total {reader.offset} bytes")
    return model_weights


# ==========================================
#           INT8 quantization ops
# ==========================================

def quantize_input(x_float: np.ndarray, scale: float, zp: int) -> np.ndarray:
    """Quantize float input to INT8"""
    x_q = np.round(x_float / scale + zp).astype(np.int32)
    return np.clip(x_q, -128, 127).astype(np.int8)


def dequantize_output(x_int: np.ndarray, scale: float, zp: int) -> np.ndarray:
    """Dequantize INT8 output back to float"""
    return (x_int.astype(np.float32) - zp) * scale


def int8_conv1d(x_int8: np.ndarray, params: LayerParams, 
               dilation: int = 1, padding: int = 0) -> np.ndarray:
    """
    INT8 Conv1d implementation (with fused ReLU)
    
    x_int8: (batch, in_channels, seq_len), int8
    weights: (out_channels, in_channels, kernel_size), int8
    
    Quantized convolution formula:
    y_int32 = sum((x_int8 - x_zp) * (w_int8 - w_zp)) + bias_int32
    y_float = y_int32 * (input_scale * weight_scale)
    y_int8 = quantize(relu(y_float), output_scale, output_zp)
    """
    batch, in_ch, seq_len = x_int8.shape
    out_ch, _, k = params.weights_int8.shape
    
    # Add padding
    if padding > 0:
        x_padded = np.pad(x_int8, ((0, 0), (0, 0), (padding, padding)), 
                         mode='constant', constant_values=params.input_zp)
    else:
        x_padded = x_int8
    
    padded_len = x_padded.shape[2]
    
    # Compute output length
    out_len = (padded_len - dilation * (k - 1) - 1) // 1 + 1
    
    # Use int32 for calculation
    x_i32 = x_padded.astype(np.int32)
    w_i32 = params.weights_int8.astype(np.int32)
    
    # Output buffer
    y_int32 = np.zeros((batch, out_ch, out_len), dtype=np.int32)
    
    # Convolution compute
    for b in range(batch):
        for oc in range(out_ch):
            w_zp = params.weight_zps[oc]
            for t in range(out_len):
                acc = 0
                for ic in range(in_ch):
                    for ki in range(k):
                        idx = t + ki * dilation
                        if idx < padded_len:
                            x_val = x_i32[b, ic, idx] - params.input_zp
                            w_val = w_i32[oc, ic, ki] - w_zp
                            acc += x_val * w_val
                
                # Add quantized bias
                acc += params.bias_int32[oc]
                y_int32[b, oc, t] = acc
    
    # Dequantize + ReLU + requantize
    y_float = np.zeros_like(y_int32, dtype=np.float32)
    for oc in range(out_ch):
        scale_factor = params.input_scale * params.weight_scales[oc]
        y_float[:, oc, :] = y_int32[:, oc, :] * scale_factor
    
    # ReLU
    y_float = np.maximum(y_float, 0)
    
    # Requantize to output scale
    y_out = np.round(y_float / params.output_scale + params.output_zp).astype(np.int32)
    y_out = np.clip(y_out, -128, 127).astype(np.int8)
    
    return y_out


def int8_conv1d_no_relu(x_int8: np.ndarray, params: LayerParams, 
                        dilation: int = 1, padding: int = 0) -> np.ndarray:
    """INT8 Conv1d implementation (without ReLU)"""
    batch, in_ch, seq_len = x_int8.shape
    out_ch, _, k = params.weights_int8.shape
    
    if padding > 0:
        x_padded = np.pad(x_int8, ((0, 0), (0, 0), (padding, padding)), 
                         mode='constant', constant_values=params.input_zp)
    else:
        x_padded = x_int8
    
    padded_len = x_padded.shape[2]
    out_len = (padded_len - dilation * (k - 1) - 1) // 1 + 1
    
    x_i32 = x_padded.astype(np.int32)
    w_i32 = params.weights_int8.astype(np.int32)
    
    y_int32 = np.zeros((batch, out_ch, out_len), dtype=np.int32)
    
    for b in range(batch):
        for oc in range(out_ch):
            w_zp = params.weight_zps[oc]
            for t in range(out_len):
                acc = 0
                for ic in range(in_ch):
                    for ki in range(k):
                        idx = t + ki * dilation
                        if idx < padded_len:
                            x_val = x_i32[b, ic, idx] - params.input_zp
                            w_val = w_i32[oc, ic, ki] - w_zp
                            acc += x_val * w_val
                acc += params.bias_int32[oc]
                y_int32[b, oc, t] = acc
    
    y_float = np.zeros_like(y_int32, dtype=np.float32)
    for oc in range(out_ch):
        scale_factor = params.input_scale * params.weight_scales[oc]
        y_float[:, oc, :] = y_int32[:, oc, :] * scale_factor
    
    y_out = np.round(y_float / params.output_scale + params.output_zp).astype(np.int32)
    y_out = np.clip(y_out, -128, 127).astype(np.int8)
    
    return y_out


def int8_linear(x_int8: np.ndarray, params: LayerParams) -> np.ndarray:
    """
    INT8 Linear implementation
    x_int8: (batch, in_features), int8
    """
    batch, in_features = x_int8.shape
    out_features = params.weights_int8.shape[0]
    
    x_i32 = x_int8.astype(np.int32)
    w_i32 = params.weights_int8.astype(np.int32)
    
    y_int32 = np.zeros((batch, out_features), dtype=np.int32)
    
    for b in range(batch):
        for oc in range(out_features):
            w_zp = params.weight_zps[oc]
            acc = 0
            for ic in range(in_features):
                x_val = x_i32[b, ic] - params.input_zp
                w_val = w_i32[oc, ic] - w_zp
                acc += x_val * w_val
            acc += params.bias_int32[oc]
            y_int32[b, oc] = acc
    
    # Dequantize
    y_float = np.zeros_like(y_int32, dtype=np.float32)
    for oc in range(out_features):
        scale_factor = params.input_scale * params.weight_scales[oc]
        y_float[:, oc] = y_int32[:, oc] * scale_factor
    
    return y_float


def int8_requantize(x_int8: np.ndarray, 
                    old_scale: float, old_zp: int,
                    new_scale: float, new_zp: int) -> np.ndarray:
    """Requantize to a new scale/zero-point"""
    x_float = (x_int8.astype(np.float32) - old_zp) * old_scale
    x_new = np.round(x_float / new_scale + new_zp).astype(np.int32)
    return np.clip(x_new, -128, 127).astype(np.int8)


def int8_add_relu(a_int8: np.ndarray, b_int8: np.ndarray,
                  a_scale: float, a_zp: int,
                  b_scale: float, b_zp: int,
                  out_scale: float, out_zp: int) -> np.ndarray:
    """INT8 residual add + ReLU"""
    # Dequantize
    a_float = (a_int8.astype(np.float32) - a_zp) * a_scale
    b_float = (b_int8.astype(np.float32) - b_zp) * b_scale
    
    # Add + ReLU
    y_float = np.maximum(a_float + b_float, 0)
    
    # Requantize
    y_out = np.round(y_float / out_scale + out_zp).astype(np.int32)
    return np.clip(y_out, -128, 127).astype(np.int8)


# ==========================================
#           Binary weight inference
# ==========================================

class BinaryInference:
    """Run INT8 inference using the .bin weight file"""
    
    def __init__(self, weights_dict):
        self.weights = weights_dict
        self.dilations = [1, 2, 4, 8]
    
    def forward_block(self, x_int8: np.ndarray, block_idx: int,
                      current_scale: float, current_zp: int):
        """Forward pass for a single block"""
        block = self.weights['blocks'][block_idx]
        dilation = self.dilations[block_idx]
        padding = (KERNEL_SIZE - 1) * dilation
        
        conv1_params = block['conv1']
        conv2_params = block['conv2']
        
        # Conv1 + ReLU (with chomp by using only the first seq_len outputs)
        out = int8_conv1d(x_int8, conv1_params, dilation=dilation, padding=padding)
        seq_len = x_int8.shape[2]
        out = out[:, :, :seq_len]  # Chomp
        
        # Conv2 + ReLU (with chomp)
        out = int8_conv1d(out, conv2_params, dilation=dilation, padding=padding)
        out = out[:, :, :seq_len]  # Chomp
        
        # Residual connection
        if block['downsample'] is not None:
            ds_params = block['downsample']
            residual = int8_conv1d_no_relu(x_int8, ds_params, dilation=1, padding=0)
        else:
            residual = x_int8
        
        # Add + ReLU
        out_int8 = int8_add_relu(
            residual, out,
            block['downsample'].output_scale if block['downsample'] else current_scale,
            block['downsample'].output_zp if block['downsample'] else current_zp,
            conv2_params.output_scale, conv2_params.output_zp,
            block['block_out_scale'], block['block_out_zp']
        )
        
        return out_int8, block['block_out_scale'], block['block_out_zp']
    
    def forward(self, x_float: np.ndarray) -> np.ndarray:
        """
        Full forward pass.
        x_float: (batch, seq_len, channels), float32 [0, 1]
        """
        # Transpose to (batch, channels, seq_len)
        x = x_float.transpose(0, 2, 1)
        
        # Get initial quantization parameters (from first conv1 input params)
        first_conv1 = self.weights['blocks'][0]['conv1']
        input_scale = first_conv1.input_scale
        input_zp = first_conv1.input_zp
        
        # Quantize input
        x_int8 = quantize_input(x, input_scale, input_zp)
        
        current_scale = input_scale
        current_zp = input_zp
        
        # Pass through each block
        for i in range(4):
            x_int8, current_scale, current_zp = self.forward_block(
                x_int8, i, current_scale, current_zp
            )
        
        # Take the last timestep
        x_last = x_int8[:, :, -1]  # (batch, channels)
        
        # Head (Linear)
        output = int8_linear(x_last, self.weights['head'])
        
        return output


# ==========================================
#           PyTorch .pth inference
# ==========================================

def load_pth_model(filepath: str) -> nn.Module:
    """Load the PyTorch PTQ quantized model"""
    model = TinyTCNInt8()
    model.fuse_model()
    model.qconfig = tq.get_default_qconfig("fbgemm")
    tq.prepare(model, inplace=True)
    model.eval()
    tq.convert(model, inplace=True)
    
    state_dict = torch.load(filepath, map_location='cpu', weights_only=False)
    model.load_state_dict(state_dict)
    model.eval()
    
    return model


def pth_inference(model: nn.Module, x: np.ndarray) -> np.ndarray:
    """Run inference with the PyTorch model"""
    x_tensor = torch.from_numpy(x).float()
    with torch.no_grad():
        output = model(x_tensor)
    return output.numpy()


# ==========================================
#           Main
# ==========================================

def main():
    print("=" * 60)
    print("TinyTCN INT8 inference comparison test (PTQ version)")
    print("=" * 60)
    
    # 1. Load test data
    print("\n[1] Loading test data...")
    data = np.load(DATA_PATH)
    X = data['X']  # uint8
    y = data['y']
    
    # Select a subset for testing
    N_TEST = 100
    X_test = X[:N_TEST].astype(np.float32) / 255.0  # normalize to [0, 1]
    y_test = y[:N_TEST]
    
    print(f"   Test samples: {N_TEST}")
    print(f"   Input shape: {X_test.shape}")  # (N, 50, 9)
    
    # 2. Load PyTorch quantized model (.pth)
    print("\n[2] Loading PyTorch quantized model (.pth)...")
    try:
        pth_model = load_pth_model(PTH_PATH)
        print("   Loaded successfully!")
    except Exception as e:
        print(f"   Failed to load: {e}")
        pth_model = None
    
    # 3. Load binary weights (.bin)
    print("\n[3] Loading binary weights (.bin)...")
    try:
        bin_weights = load_binary_weights(BIN_PATH)
        bin_inference = BinaryInference(bin_weights)
        print("   Loaded successfully!")
    except Exception as e:
        print(f"   Failed to load: {e}")
        bin_inference = None
    
    # 4. Run inference comparison
    print("\n[4] Inference comparison...")
    print("-" * 60)
    
    if pth_model is not None:
        print("\n>>> PyTorch (.pth) inference:")
        pth_outputs = pth_inference(pth_model, X_test)
        pth_probs = 1 / (1 + np.exp(-pth_outputs))  # Sigmoid
        pth_preds = (pth_probs > 0.5).astype(np.float32).flatten()
        
        pth_acc = (pth_preds == y_test).mean()
        print(f"    Accuracy: {pth_acc:.4f}")
        print(f"    Output range: [{pth_outputs.min():.4f}, {pth_outputs.max():.4f}]")
        print(f"    Output mean: {pth_outputs.mean():.4f}")
    
    if bin_inference is not None:
        print("\n>>> Binary (.bin) INT8 inference:")
        bin_outputs = bin_inference.forward(X_test)
        bin_probs = 1 / (1 + np.exp(-bin_outputs))  # Sigmoid
        bin_preds = (bin_probs > 0.5).astype(np.float32).flatten()
        
        bin_acc = (bin_preds == y_test).mean()
        print(f"    Accuracy: {bin_acc:.4f}")
        print(f"    Output range: [{bin_outputs.min():.4f}, {bin_outputs.max():.4f}]")
        print(f"    Output mean: {bin_outputs.mean():.4f}")
    
    # 5. Compare the two inference paths
    if pth_model is not None and bin_inference is not None:
        print("\n[5] Comparison of the two inference paths:")
        print("-" * 60)
        
        # Output differences
        output_diff = np.abs(pth_outputs.flatten() - bin_outputs.flatten())
        print(f"    Output absolute difference:")
        print(f"      - Max: {output_diff.max():.6f}")
        print(f"      - Mean: {output_diff.mean():.6f}")
        print(f"      - Median: {np.median(output_diff):.6f}")
        
        # Prediction agreement
        pred_match = (pth_preds == bin_preds).mean()
        print(f"\n    Prediction agreement: {pred_match:.4f} ({int(pred_match * N_TEST)}/{N_TEST})")
        
        # Accuracy delta
        acc_diff = abs(pth_acc - bin_acc)
        print(f"    Accuracy delta: {acc_diff:.4f}")
        
        # Correlation coefficient
        correlation = np.corrcoef(pth_outputs.flatten(), bin_outputs.flatten())[0, 1]
        print(f"    Output correlation: {correlation:.6f}")
        
        # Judge whether the difference is acceptable
        print("\n" + "=" * 60)
        if output_diff.max() < 0.5 and pred_match > 0.95:
            print("✓ Conclusion: Results are highly consistent; .bin file is good to use.")
        elif output_diff.max() < 1.0 and pred_match > 0.9:
            print("△ Conclusion: Minor differences; likely due to quantization precision loss.")
        else:
            print("✗ Conclusion: Large differences; please check the weight export/load logic.")
    
    # 6. Detailed per-sample comparison
    print("\n[6] Detailed per-sample comparison (first 5 samples):")
    print("-" * 60)
    print(f"{'Sample':<8} {'Label':<8} {'PTH Output':<15} {'BIN Output':<15} {'Diff':<10}")
    print("-" * 60)
    
    for i in range(min(5, N_TEST)):
        pth_out = pth_outputs[i, 0] if pth_model else float('nan')
        bin_out = bin_outputs[i, 0] if bin_inference else float('nan')
        diff = abs(pth_out - bin_out) if (pth_model and bin_inference) else float('nan')
        print(f"{i:<8} {y_test[i]:<8.0f} {pth_out:<15.6f} {bin_out:<15.6f} {diff:<10.6f}")
    
    print("\n" + "=" * 60)
    print("Inference comparison complete!")


if __name__ == "__main__":
    main()

File: TCN/CModel/inference.c
/**
 * @file inference.c
 * @brief TinyTCN INT8 inference implementation - ASIC Golden Model (PTQ version)
 */

#include "inference.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>

/* ========================================
 *          Buffer management
 * ======================================== */

int inference_buffer_init(InferenceBuffer *buf) {
    buf->buffer_size = MAX_CHANNELS * WINDOW_SIZE;
    
    buf->buffer_a = (int8_t *)calloc(buf->buffer_size, sizeof(int8_t));
    buf->buffer_b = (int8_t *)calloc(buf->buffer_size, sizeof(int8_t));
    buf->temp1 = (int8_t *)calloc(buf->buffer_size, sizeof(int8_t));
    buf->temp2 = (int8_t *)calloc(buf->buffer_size, sizeof(int8_t));
    
    if (!buf->buffer_a || !buf->buffer_b || !buf->temp1 || !buf->temp2) {
        inference_buffer_free(buf);
        return -1;
    }
    
    buf->current = buf->buffer_a;
    buf->next = buf->buffer_b;
    
    return 0;
}

void inference_buffer_free(InferenceBuffer *buf) {
    if (buf->buffer_a) { free(buf->buffer_a); buf->buffer_a = NULL; }
    if (buf->buffer_b) { free(buf->buffer_b); buf->buffer_b = NULL; }
    if (buf->temp1) { free(buf->temp1); buf->temp1 = NULL; }
    if (buf->temp2) { free(buf->temp2); buf->temp2 = NULL; }
}

static void swap_buffers(InferenceBuffer *buf) {
    int8_t *tmp = buf->current;
    buf->current = buf->next;
    buf->next = tmp;
}

/* ========================================
 *          Quantization basics
 * ======================================== */

int8_t quantize_int8(float val, float scale, int32_t zp) {
    int32_t q = (int32_t)roundf(val / scale) + zp;
    if (q < -128) q = -128;
    if (q > 127) q = 127;
    return (int8_t)q;
}

float dequantize_int8(int8_t val, float scale, int32_t zp) {
    return ((float)val - (float)zp) * scale;
}

/* ========================================
 *          INT8 Conv1D
 * ======================================== */

/**
 * @brief INT8 Conv1D core implementation
 * 
 * Input layout: [channels, seq_len] (CHW format)
 * Weight layout: [out_ch, in_ch, kernel_size]
 */
int int8_conv1d(int8_t *output,
                const int8_t *input,
                const LayerParams *params,
                int in_len,
                int dilation,
                int apply_relu) {
    
    int out_ch = params->out_channels;
    int in_ch = params->in_channels;
    int k = params->kernel_size;
    
    // Causal padding (left-side padding)
    int padding = (k - 1) * dilation;
    (void)padding;  // used for calculations below
    
    // Output length (chomp keeps it same as input)
    int out_len = in_len;
    
    int32_t x_zp = params->input_zp;
    
    // Loop over output channels
    for (int oc = 0; oc < out_ch; oc++) {
        int32_t w_zp = params->weight_zps[oc];
        float w_scale = params->weight_scales[oc];
        float scale_factor = params->input_scale * w_scale;
        
        // Loop over timesteps
        for (int t = 0; t < out_len; t++) {
            int32_t acc = 0;
            
            // Kernel traversal
            for (int ic = 0; ic < in_ch; ic++) {
                for (int ki = 0; ki < k; ki++) {
                    // Compute input index (consider padding and dilation)
                    int t_in = t + ki * dilation;  // relative to padded input
                    
                    int32_t x_val;
                    if (t_in < padding) {
                        // In padding region, use zero point
                        x_val = x_zp;
                    } else {
                        int actual_idx = t_in - padding;
                        if (actual_idx < in_len) {
                            x_val = (int32_t)input[ic * in_len + actual_idx];
                        } else {
                            x_val = x_zp;
                        }
                    }
                    
                    // Fetch weight
                    int w_idx = oc * (in_ch * k) + ic * k + ki;
                    int32_t w_val = (int32_t)params->weights_int8[w_idx];
                    
                    // Accumulate: (x - x_zp) * (w - w_zp)
                    acc += (x_val - x_zp) * (w_val - w_zp);
                }
            }
            
            // Add quantized bias
            acc += params->bias_int32[oc];
            
            // Dequantize to float
            float y_float = (float)acc * scale_factor;
            
            // ReLU (if needed)
            if (apply_relu && y_float < 0) {
                y_float = 0.0f;
            }
            
            // Requantize to output scale
            output[oc * out_len + t] = quantize_int8(y_float, 
                                                      params->output_scale, 
                                                      params->output_zp);
        }
    }
    
    return out_len;
}

/* ========================================
 *          INT8 residual add
 * ======================================== */

void int8_add_relu(int8_t *output,
                   const int8_t *a, const int8_t *b,
                   int channels, int seq_len,
                   float a_scale, int32_t a_zp,
                   float b_scale, int32_t b_zp,
                   float out_scale, int32_t out_zp) {
    
    for (int c = 0; c < channels; c++) {
        for (int t = 0; t < seq_len; t++) {
            int idx = c * seq_len + t;
            
            // Dequantize
            float a_float = dequantize_int8(a[idx], a_scale, a_zp);
            float b_float = dequantize_int8(b[idx], b_scale, b_zp);
            
            // Add + ReLU
            float sum = a_float + b_float;
            if (sum < 0) sum = 0.0f;
            
            // Requantize
            output[idx] = quantize_int8(sum, out_scale, out_zp);
        }
    }
}

/* ========================================
 *          INT8 Linear
 * ======================================== */

float int8_linear(const int8_t *input, const LayerParams *params) {
    int in_features = params->in_channels;  // Linear: in_channels = in_features
    int32_t x_zp = params->input_zp;
    
    // Only one output (out_channels = 1)
    int32_t w_zp = params->weight_zps[0];
    float w_scale = params->weight_scales[0];
    
    int32_t acc = 0;
    
    for (int i = 0; i < in_features; i++) {
        int32_t x_val = (int32_t)input[i];
        int32_t w_val = (int32_t)params->weights_int8[i];
        
        acc += (x_val - x_zp) * (w_val - w_zp);
    }
    
    // Add bias
    acc += params->bias_int32[0];
    
    // Dequantize to float (no requantization; direct output)
    float scale_factor = params->input_scale * w_scale;
    return (float)acc * scale_factor;
}

/* ========================================
 *          TCN block forward
 * ======================================== */

static void forward_block(const BlockParams *block,
                          int8_t *output,
                          const int8_t *input,
                          int8_t *temp_conv1,   // for conv1 output
                          int8_t *temp_conv2,   // for conv2 output
                          int out_ch,
                          int seq_len,
                          float *current_scale,
                          int32_t *current_zp) {
    
    // Conv1 + ReLU
    int8_conv1d(temp_conv1, input, &block->conv1, seq_len, 
                block->dilation, 1);  // apply_relu = 1
    
    // Conv2 + ReLU
    int8_conv1d(temp_conv2, temp_conv1, &block->conv2, seq_len,
                block->dilation, 1);  // apply_relu = 1
    
    // Residual branch
    const int8_t *residual;
    float res_scale;
    int32_t res_zp;
    
    if (block->has_downsample) {
        // Downsample (1x1 conv, no ReLU) - reuse temp_conv1 for output (conv1 already used)
        int8_conv1d(temp_conv1, input, &block->downsample, seq_len,
                    1, 0);  // dilation=1, no relu
        residual = temp_conv1;
        res_scale = block->downsample.output_scale;
        res_zp = block->downsample.output_zp;
    } else {
        residual = input;  // use input directly
        res_scale = *current_scale;
        res_zp = *current_zp;
    }
    
    // Residual add + ReLU
    int8_add_relu(output,
                  residual, temp_conv2,
                  out_ch, seq_len,
                  res_scale, res_zp,
                  block->conv2.output_scale, block->conv2.output_zp,
                  block->block_out_scale, block->block_out_zp);
    
    // Update current quantization params
    *current_scale = block->block_out_scale;
    *current_zp = block->block_out_zp;
}

/* ========================================
 *          Full inference
 * ======================================== */

float inference_forward(const TinyTCNModel *model, 
                        InferenceBuffer *buf,
                        const uint8_t *input) {
    
    int seq_len = WINDOW_SIZE;
    
    // 1. Input quantization: uint8 [0-255] -> float [0-1] -> int8
    // Layout conversion: [T, C] -> [C, T]
    for (int c = 0; c < NUM_INPUTS; c++) {
        for (int t = 0; t < seq_len; t++) {
            // Raw input is uint8 ADC code, normalize to [0, 1]
            float val = (float)input[t * NUM_INPUTS + c] / 255.0f;
            buf->current[c * seq_len + t] = quantize_int8(val, 
                                                          model->input_scale, 
                                                          model->input_zp);
        }
    }
    
    float current_scale = model->input_scale;
    int32_t current_zp = model->input_zp;
    
    // Output channel configuration
    int out_channels[NUM_BLOCKS] = {CH_BLOCK0, CH_BLOCK1, CH_BLOCK2, CH_BLOCK3};
    
    // 2. Forward pass through each block
    for (int i = 0; i < NUM_BLOCKS; i++) {
        forward_block(&model->blocks[i],
                      buf->next,        // output
                      buf->current,     // input
                      buf->temp1,       // temp for conv1
                      buf->temp2,       // temp for conv2
                      out_channels[i],
                      seq_len,
                      &current_scale, &current_zp);
        swap_buffers(buf);
    }
    
    // 3. Take the last timestep
    int8_t last_timestep[CH_BLOCK3];
    for (int c = 0; c < CH_BLOCK3; c++) {
        last_timestep[c] = buf->current[c * seq_len + (seq_len - 1)];
    }
    
    // 4. Head (Linear)
    return int8_linear(last_timestep, &model->head);
}

void inference_batch(const TinyTCNModel *model,
                     InferenceBuffer *buf,
                     const uint8_t *inputs,
                     float *outputs,
                     int batch_size) {
    
    int sample_size = WINDOW_SIZE * NUM_INPUTS;
    
    for (int i = 0; i < batch_size; i++) {
        outputs[i] = inference_forward(model, buf, inputs + i * sample_size);
    }
}

